File: examples/python/example03.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
@author: sorenaf
"""

import embanded
import scipy.io
import matplotlib.pyplot as plt
import numpy as np

# Load data from the 'example03.mat' file
data = scipy.io.loadmat('example03.mat')


# Split the data into predictor groups (F) and target variable (y)
X = np.concatenate(data['F'].ravel(),axis=1)
num_features = X.shape[1]
F = []
for j in range(X.shape[1]):
    F.append(X[:,j][:,None])
y = data['y']


# Check if the predictors and target variable are centered
assert np.isclose(np.concatenate(F,axis=1).mean(axis=0),0).all(), "The predictors have been centered, please see the MATLAB simulations"
assert np.isclose(y.mean(axis=0),0).all(), "The target variable has been centered, please see the MATLAB simulations"


# Create a grid of subplots for plotting results
f,ax = plt.subplots(2,4,sharex=True,figsize=(15,7.5))

for k, param in enumerate([1e-4,1e-3,1e-2,1e-1]):

    clf = embanded.EMBanded(num_features=num_features,remove_intercept=False,max_iterations=200,
                                tau =param,
                                phi=param,
                                eta=param,
                                kappa=param,
                                multi_dimensional=False)
    
    # Fit the model
    summary = clf.fit(F,y)
    
    
    # Plot the estimated weights for this parameter
    ax[0,k].plot(clf.W,'-k')
    ax[0,k].set_title(r'$\eta=\phi=\kappa=\tau=%0.1e$'%param)
    
    
    # Check if the estimated weights match the provided data
    assert np.isclose(data['W_estimated'][0,k],clf.W).all(), 'The estimated weights are not matching'



    # Import the Ridge regression model from scikit-learn
    import sklearn.linear_model
    
    # Initialize the Ridge regression model with the specified alpha
    clf_ridge = sklearn.linear_model.Ridge(alpha=1./param,fit_intercept=False,solver='cholesky',copy_X=True)
    
    # Concatenate predictors (F) and fit the Ridge regression model
    X = np.concatenate(F,axis=1)

    # Plot the estimated weights
    ax[1,k].plot(clf_ridge.fit(X,y).coef_.ravel(),'-r')
    ax[1,k].set_title(r'$\alpha=%0.1e$'%(1./param))
    
    # Check if the estimated weights match the provided data
    assert np.isclose(clf_ridge.fit(X,y).coef_.ravel(),data['W_ridge'][0,k].ravel()).all()

File: examples/matlab/example03.m
clear; clc; close all; rng('default')
assert(~isempty(strfind(pwd, 'embanded/examples/matlab')), 'Please change folder to the examples directory')
run(strrep(pwd,'examples/matlab','setup.m'))

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Simulate data.

% Generate some training data
num_obs = 1024;
num_dim = 512;

% Define covariane matrix of regressors
C = exp(-0.3*([0:num_dim-1]-[0:num_dim-1]').^2);

% Simulate predictors
X = zscore(mvnrnd(zeros(1,num_dim),C,num_obs));

% Simulate weights (most weights will be equal to zero)
W = randn(num_dim,1).*(rand(num_dim,1)>0.95);

% Simulate noise term
N = zscore(mvnrnd(0,1,num_obs));

[A,B] = scaledata(X,W,N,0);

W = W*A;
N = N*B;

% Simulate the response
y = X*W + N;

% Display stuff for the interested
fprintf('\n Standard deviation of y: %0.2f', std(y))
fprintf('\n Target-to-noise ratio: %0.2f dB',10*log10(mean((X*W).^2)/mean(N.^2)))


% Split the data into predictor groups (F). This example assigns distinct 
% hyperparameters to each predictor.
F = cell(1,num_dim);
for f = 1 : num_dim
    F{1,f} = X(:,f);
end

% Sweep through the following parameter values
param_values =  [1e-4,1e-3,1e-2,1e-1];
W_estimated = cell(1,4);
W_ridge = cell(1,4);

for pp = 1 : 4
    param_val = param_values(pp);
    
    % Estimate with EM-banded
    opts = struct;
    opts.max_iterations = 200;
    opts.show_progress = false;
    opts.kappa = param_val;
    opts.phi = param_val;
    opts.tau = param_val;
    opts.eta = param_val;
    [W_estimated{pp},summary] = embanded(F,y,opts);
    
   
    
    % Plot the estimated weights
    subplot(4,5,[pp,pp+5])
    plot(W_estimated{pp},'-k','LineWidth',1)
    xlim([1,num_dim]);
    ylim([-1,1]*max(abs(W(:))))
    title(['\eta=\phi=\tau=\kappa=',num2str(param_val)],'FontWeight','Normal','Fontsize',12)
    ylabel('Weights (a.u.)','Fontsize',12)
    
    % Estimate with Ridge
    L = 1./param_values(pp)*eye(size(X,2));
    W_ridge{pp} = (X'*X+L)\(X'*y);
    
    
    % Plot the estimated weights
    subplot(4,5,[pp+10,pp+15])
    plot(W_ridge{pp},'-k','LineWidth',1)
    ylim([-1,1]*max(abs(W_ridge{pp}(:))))
    hold on
    xlim([1,length(W)]);
    title(['1/\alpha=',num2str(param_val)],'FontWeight','Normal','Fontsize',12)
    ylabel('Weights (a.u.)','Fontsize',12)

    
    
end

subplot(4,5,[10 15])
plot(W,'-k','LineWidth',1)
ylim([-1,1]*max(abs(W(:))))
xlim([1,length(W)]);
title('Target','FontWeight','Normal','Fontsize',12)
ylabel('Weights (a.u.)','Fontsize',12)


fig = get(groot,'CurrentFigure');
fig.Position = [100 100 1500 400];

% The following command was used to store the file "example03.mat":
% save(strrep(pwd,'matlab','python/example03.mat'),'W_estimated','F','y','W','W_ridge')
